### PR TITLE
[WIP] Rhapsodies of Vana'diel 1-1 to 1-18

### PR DESCRIPTION
**Pull request by zach2good**
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](https://github.com/project-topaz/topaz/blob/master/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

**TODO:**
Clean up commit history
Verify EXP bonus is suitable
Test Gilgamesh's sub job letter workflow front to back
Address the rest of the PR comments

**Doing in later PRs:**
Balance Ophiotaurus and Siren fights
Other Rhapsody in .* bonuses

**DONE:**
Structure of RoV 1-1 to 1-18 ✔️ 
Handling of Mhuara/Selbina paths ✔️ 
CS's in all starting zones ✔️ 
Undo auto-format in status.lua etc. ✔️ 
Triple check my port didn't squish anyone else's changes ✔️ 
Push to my server and make available for testing ✔️ 
Replace things that were squished in topaz port ✔️ 
KeyItem based subjob handling ✔️ 
Testing: Ring My Bell: First character lockout and use of `rhapsodies.lua`: ✔️ 

**Testing Notes:**
All RoV missions from 1-1 to 1-18 are scripted.
You will need to set ENABLE_ROV to 1
Later on you will need rank 3
Towards the end you will need to set you nation mission to 16 (defeated shadowlord)

**Mhaura path:**
!setplayervar <name> RhapsodiesStatus 2

**Selbina path:**
!setplayervar <name> RhapsodiesStatus 1

**Items for SET_FREE:**
Ekokoko:
!additem 9083 3

Abelard
!additem 9082 3

**Ophiotaurus spawn:**
!setmission 13 20

### CODE REVIEW

File: scripts/globals/rhapsodies.lua
Comment (ibm2431) [inline scripts/globals/rhapsodies.lua]: When the set constructor PR gets merged, you'll be able to do some wonderful things:
```lua
tpz.rhapsodies.unavailability =
{
    [tpz.rhapsodies.character.PRISHE] = set{
       tpz.mission.id.cop.THE_RITES_OF_LIFE,
       tpz.mission.id.cop.BELOW_THE_ARKS,
    },
    [tpz.rhapsodies.character.TENZEN] = set{
       tpz.mission.id.cop.THE_RITES_OF_LIFE,
       tpz.mission.id.cop.BELOW_THE_ARKS,
    }
}
```

and then... (continued next comment)

File: scripts/globals/rhapsodies.lua
Comment (ibm2431) [inline scripts/globals/rhapsodies.lua]:
```lua
tpz.rhapsodies.requiredCharacters =
{
    [tpz.mission.id.rov.NUMBERING_DAYS] = {
        tpz.rhapsodies.character.PRISHE, 
        tpz.rhapsodies.character.TENZEN
    }
}

tpz.rhapsodies.charactersAvailable = function(player)
    local rov_mission = player:getCurrentMission(ROV)
    for _, char in pairs(tpz.rhapsodies.requiredCharacters[rov_mission]) do
        local expansion_mission = player:getCurrentMission(tpz.rhapsodies.expansion[char])
        if tpz.rhapsodies.unavailability[char][expansion_mission] then
            return false
        end
    end
    return true
end
```

And all the individual `characterAvailable` functions can be cut. When you expect a character check for a mission, you just call the the `tpz.rhapsodies.charactersAvailable(player)` function without having to worry about `and`-ing multiple character checks together.

But you'll have to define a character->expansion lookup table.

File: scripts/globals/rhapsodies.lua
Comment (ibm2431) [inline scripts/globals/rhapsodies.lua]: Add a character->expansion lookup table here, after defining your character enums:
```lua
tpz.rhapsodies.expansion = 
{
    [tpz.rhapsodies.character.PRISHE] = COP,
    [tpz.rhapsodies.character.APHMAU] = TOAU,
}
```

File: scripts/zones/Mhaura/npcs/Tonasav.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/npcs/Tonasav.lua:28]: Putting a note here to double-check later where exactly retail drops you.

File: scripts/zones/Selbina/npcs/Pacomart.lua
Comment (ibm2431) [inline scripts/zones/Selbina/npcs/Pacomart.lua:28]: Putting a note here to double-check later where exactly retail drops you.

File: scripts/zones/Selbina/npcs/Pacomart.lua
Comment (ibm2431) [inline scripts/zones/Selbina/npcs/Pacomart.lua]: These NPCs will still transport you after you complete the mission.

I'll have to post it to the Captures server, but I do have a capture from last year in which I tested it immediately after getting the Rhapsody in White.

File: scripts/zones/Mhaura/npcs/Tonasav.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/npcs/Tonasav.lua]: Same here re: transporting after the mission

File: scripts/zones/Mhaura/npcs/Ekokoko.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/npcs/Ekokoko.lua]: Remember to use the boolean returned by `npcUtil.giveItem` to make sure a player was able to receive an inventory item before you take items from them and complete a quest~

File: scripts/zones/Mhaura/Zone.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/Zone.lua]: Mhaura's path needs to be something other than zero, or else the player will be swapping paths if they zone into Selbina~!

File: scripts/zones/Mhaura/npcs/Ekokoko.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/npcs/Ekokoko.lua]: And then check for the new Mhaura path variable here

File: scripts/zones/Mhaura/npcs/Vera.lua
Comment (ibm2431) [inline scripts/zones/Mhaura/npcs/Vera.lua]: This isn't your fault, but this var is missing a `local`

File: scripts/zones/Selbina/npcs/Abelard.lua
Comment (ibm2431) [inline scripts/zones/Selbina/npcs/Abelard.lua]: Be sure that the player got their item before advancing here too

File: scripts/zones/Selbina/npcs/Isacio.lua
Comment (ibm2431) [inline scripts/zones/Selbina/npcs/Isacio.lua]: Also not your fault, but the `local` is missing here too

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua:6]: Only going to make the following comments on _this_ telepoint, but they apply to the other two as well:

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua]: As a general rule, ROV cutscenes overrule all expansion cutscenes. Unfortunately, I don't have a capture testing the ROV/COP interaction here - in the only video I have this early in ROV, I talked to it twice and got the "million pieces" message, so I must not have flagged Below the Arks yet.

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua]: The requirement here is _rank_ 3, not fame level 3. 😉

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua:38]: You'll need to check for the character's ROV here too, because... (see next)

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua:63]: If you trigger the scene with a full inventory, you'll get the full scene, but the mission will not advance because it failed to give you Lion's Cipher. However, unlike other CS events which will play out all over again if you re-trigger the event, the Shattered Telepoint remembers you already viewed the event and just tries to give you the cipher again. When you successfully acquire the Cipher, _then_ you'll advance to the next mission.

You can see this behavior in "Rhapsodies: A Land After Time (1-11) (Full Inventory)".

The behavior has also held true for all other Ciphers ROV has tried to give me when I've remembered to keep full inventory.

So what'll need to be done is something like:
```lua
if npcUtil.giveItem then
    -- advance
else
    player:setCharVar("rhapsodies", 1)
end
```

And then... (next)

File: scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua
Comment (ibm2431) [inline scripts/zones/Konschtat_Highlands/npcs/Shattered_Telepoint.lua:40]: ..back up in the trigger, check for the same mission and the "saw already" var, and then try giving the item again directly (and advancing mission on success) without replaying the CS.

File: scripts/zones/Norg/npcs/_700.lua
Comment (ibm2431) [inline scripts/zones/Norg/npcs/_700.lua]: So while I can't show the ROV/COP interaction for the Shattered Telepoint, I _can_ show definetively that _all_ of _these_ ROV cutscenes take priority over the ZM 2, in that it's possible to reach ROV 2-19 without the door flagging the ZM 2 CS.

See: "Rhapsodies: ROV 2-18 ~ ROV 2-24 ("Lowest", Lunette) + ZM 2 + COP 4-1" at 23m40s.

I accidentally proved it twice; once with the character already on ROV 2-19, and then when I _tried_ to flag the ZM2 CS with the second character, but got two ROV CSes before finally getting the ZM2 CS because _that_ character was on ROV 1-17 (Volto Oscuro).

It's outside the scope of this PR, but for future reference for anyone interested - this holds true until ROV 2-23. At that point, even if you have a ROV CS at the door, it will choose ZM 2 _first_, and then you'll be blocked on ROV 2-23/2-24 until you clear ZM 3. (See 54m of the same video.)

File: scripts/zones/Qufim_Island/npcs/Undulating_Confluence.lua
Comment (ibm2431) [inline scripts/zones/Qufim_Island/npcs/Undulating_Confluence.lua]: Confrontation still being used here~

File: src/map/utils/charutils.cpp
Comment (ibm2431) [inline src/map/utils/charutils.cpp]: This bonus appears to be a static exp value being added to the base; instead of adding to the exp multiplier down below:

File: src/map/utils/charutils.cpp
Comment (ibm2431) [inline src/map/utils/charutils.cpp]: So if I'm reading this right, the ROV KI would only add 30 exp, instead of multiplying by 30%.

**Review by ibm2431 — CHANGES_REQUESTED** [review]
First, thanks as always, Zach!

As we discussed on Discord, if you won't be able to work on this for awhile I'm okay with squash-merging this to a new ROV branch, and then someone else can pick up on the comments I have here~!

I'll admit I didn't look too closely at the mobskills since there's others that need implementing. We'll have to work on that too before merging any new ROV branch into master/canary.

File: scripts/globals/confrontation.lua
Comment (TeoTwawki) [inline scripts/globals/confrontation.lua]: the heck is all this?? I was able to tie it all to player and mob in the core the same way bcnm is tied. was much shorter  because I was able to reuse existing code.

(maybe its too early in the morning but I don't follow the purpose of doing all of this lua side)

File: scripts/zones/Mhaura/npcs/Vera.lua
Comment (TeoTwawki) [reply scripts/zones/Mhaura/npcs/Vera.lua]: she's a vera old lady, she forgot to stop being a global var forever ago

File: src/map/utils/charutils.cpp
Comment (TeoTwawki) [reply src/map/utils/charutils.cpp]:
```cpp
        auto RoVbonus = 0;
        for (auto i = 2884; i <= 2892; ++i) // RHAPSODY KI are sequential, so start at WHITE and end at OCHRE
        {
            if (hasKeyItem(PChar, i))
            {
                RoVbonus += 30;
```
```cpp
        bonus += (int32)(exp * (PChar->getMod(Mod::EXP_BONUS) + RoVbonus / 100.0f));
```
should fix that I think

File: src/map/zone.cpp
Comment (TeoTwawki) [inline src/map/zone.cpp]: shouldn't removal on zone already happen by way of effect flags?

File: scripts/globals/confrontation.lua
Comment (TeoTwawki) [reply scripts/globals/confrontation.lua]: yo dude i didn't mean to have you delete all your work - I dunno what you mean to do wid it yet

**Review by ibm2431 — APPROVED** [review]
Approved for merging into ROV branch; we can address mobskills and other small things there before a merge into canary

**Pull request merged**
